pgmpy/factors/Factor.py

```diff
@@ -780,10 +780,12 @@ def __ne__(self, other):
         return not self.__eq__(other)
 
     def __hash__(self):
-        variables = sorted(self.variables)
+        variable_hashes = [hash(variable) for variable in self.variables]
+        sorted_var_hashes = list(sorted(variable_hashes))
         phi = self.copy()
         for axis in range(phi.values.ndim):
-            exchange_index = phi.variables.index(variables[axis])
+            exchange_index = variable_hashes.index(sorted_var_hashes[axis])
+            variable_hashes[axis], variable_hashes[exchange_index] = (variable_hashes[exchange_index], variable_hashes[axis])
             phi.variables[axis], phi.variables[exchange_index] = (phi.variables[exchange_index],
```

Comment:
@abhi95 Because we need to use the hash values of the objects for computing the final hash. Right now str (in line 794) would return different value for two different objects having the same attributes.